File: src/rendering/Camera.cpp
#include "poorcraft/rendering/Camera.h"

#include <glm/gtc/matrix_transform.hpp>
#include <glm/gtc/type_ptr.hpp>

#include <cmath>

#include "poorcraft/core/Logger.h"

namespace PoorCraft {

namespace {
constexpr float DEFAULT_FOV = 90.0f;
constexpr float DEFAULT_ASPECT = 16.0f / 9.0f;
constexpr float DEFAULT_NEAR = 0.1f;
constexpr float DEFAULT_FAR = 1000.0f;
}

Camera::Camera(CameraType type,
               const glm::vec3& position,
               const glm::vec3& target,
               const glm::vec3& up)
    : m_Type(type),
      m_Position(position),
      m_Target(target),
      m_Up(glm::normalize(up)),
      m_ViewMatrix(1.0f),
      m_ProjectionMatrix(1.0f),
      m_FOV(DEFAULT_FOV),
      m_AspectRatio(DEFAULT_ASPECT),
      m_NearPlane(DEFAULT_NEAR),
      m_FarPlane(DEFAULT_FAR) {
    updateViewMatrix();
    if (m_Type == CameraType::PERSPECTIVE) {
        updateProjectionMatrix(m_FOV, m_AspectRatio, m_NearPlane, m_FarPlane);
    }
}

void Camera::updateViewMatrix() {
    m_ViewMatrix = glm::lookAt(m_Position, m_Target, m_Up);
    PC_TRACE("Camera view matrix updated");
}

void Camera::updateProjectionMatrix(float fov, float aspect, float nearPlane, float farPlane) {
    m_Type = CameraType::PERSPECTIVE;
    m_FOV = fov;
    m_AspectRatio = aspect;
    m_NearPlane = nearPlane;
    m_FarPlane = farPlane;

    m_ProjectionMatrix = glm::perspective(glm::radians(fov), aspect, nearPlane, farPlane);
    PC_TRACE("Camera perspective projection updated");
}

void Camera::updateProjectionMatrix(float left, float right, float bottom, float top, float nearPlane, float farPlane) {
    m_Type = CameraType::ORTHOGRAPHIC;
    m_ProjectionMatrix = glm::ortho(left, right, bottom, top, nearPlane, farPlane);
    m_NearPlane = nearPlane;
    m_FarPlane = farPlane;
    PC_TRACE("Camera orthographic projection updated");
}

void Camera::setPosition(const glm::vec3& position) {
    m_Position = position;
    updateViewMatrix();
}

void Camera::setTarget(const glm::vec3& target) {
    m_Target = target;
    updateViewMatrix();
}

void Camera::setUp(const glm::vec3& up) {
    m_Up = glm::normalize(up);
    updateViewMatrix();
}

void Camera::lookAt(const glm::vec3& target) {
    setTarget(target);
}

void Camera::move(const glm::vec3& offset) {
    m_Position += offset;
    m_Target += offset;
    updateViewMatrix();
}

void Camera::rotate(float yawRadians, float pitchRadians) {
    glm::vec3 forward;
    forward.x = cosf(pitchRadians) * cosf(yawRadians);
    forward.y = sinf(pitchRadians);
    forward.z = cosf(pitchRadians) * sinf(yawRadians);
    forward = glm::normalize(forward);

    m_Target = m_Position + forward;
    updateViewMatrix();
}

const glm::mat4& Camera::getViewMatrix() const {
    return m_ViewMatrix;
}

const glm::mat4& Camera::getProjectionMatrix() const {
    return m_ProjectionMatrix;
}

glm::mat4 Camera::getViewProjectionMatrix() const {
    return m_ProjectionMatrix * m_ViewMatrix;
}

const glm::vec3& Camera::getPosition() const {
    return m_Position;
}

const glm::vec3& Camera::getTarget() const {
    return m_Target;
}

glm::vec3 Camera::getForward() const {
    return glm::normalize(m_Target - m_Position);
}

glm::vec3 Camera::getRight() const {
    return glm::normalize(glm::cross(getForward(), m_Up));
}

const glm::vec3& Camera::getUp() const {
    return m_Up;
}

float Camera::getFOV() const {
    return m_FOV;
}

float Camera::getAspectRatio() const {
    return m_AspectRatio;
}

float Camera::getNearPlane() const {
    return m_NearPlane;
}

float Camera::getFarPlane() const {
    return m_FarPlane;
}

void Camera::onWindowResize(int width, int height) {
    if (height == 0) {
        return;
    }
    m_AspectRatio = static_cast<float>(width) / static_cast<float>(height);
    if (m_Type == CameraType::PERSPECTIVE) {
        updateProjectionMatrix(m_FOV, m_AspectRatio, m_NearPlane, m_FarPlane);
    } else {
        PC_DEBUG("Camera window resize ignored for orthographic camera - call orthographic update explicitly");
    }
}

} // namespace PoorCraft

File: client/src/main/java/com/poorcraft/client/render/camera/Camera.java
package com.poorcraft.client.render.camera;

import com.poorcraft.common.Constants;
import org.joml.Matrix4f;
import org.joml.Quaternionf;
import org.joml.Vector3f;

/**
 * 3D perspective camera using JOML for matrix math.
 */
public class Camera {
    private final Vector3f position;
    private final Quaternionf rotation;
    
    private float yaw;
    private float pitch;
    
    private float fov;
    private float aspectRatio;
    private float nearPlane;
    private float farPlane;
    private boolean useReverseZ;
    
    private final Matrix4f viewMatrix;
    private final Matrix4f projectionMatrix;
    private final Matrix4f viewProjectionMatrix;
    
    private boolean viewDirty;
    private boolean projectionDirty;
    
    /**
     * Creates a new camera with default settings.
     */
    public Camera() {
        this.position = new Vector3f(0, 0, 0);
        this.rotation = new Quaternionf();
        this.yaw = 0.0f;
        this.pitch = 0.0f;
        
        this.fov = Constants.Rendering.DEFAULT_FOV;
        this.aspectRatio = 16.0f / 9.0f;
        this.nearPlane = Constants.Rendering.DEFAULT_NEAR_PLANE;
        this.farPlane = Constants.Rendering.DEFAULT_FAR_PLANE;
        this.useReverseZ = false;
        
        this.viewMatrix = new Matrix4f();
        this.projectionMatrix = new Matrix4f();
        this.viewProjectionMatrix = new Matrix4f();
        
        this.viewDirty = true;
        this.projectionDirty = true;
    }
    
    /**
     * Sets the camera position.
     * 
     * @param position The new position
     */
    public void setPosition(Vector3f position) {
        this.position.set(position);
        viewDirty = true;
    }
    
    /**
     * Sets the camera position.
     * 
     * @param x X coordinate
     * @param y Y coordinate
     * @param z Z coordinate
     */
    public void setPosition(float x, float y, float z) {
        this.position.set(x, y, z);
        viewDirty = true;
    }
    
    /**
     * Sets the camera rotation using yaw and pitch angles.
     * 
     * @param yaw Yaw angle in degrees
     * @param pitch Pitch angle in degrees
     */
    public void setRotation(float yaw, float pitch) {
        this.yaw = yaw;
        this.pitch = Math.max(-89.0f, Math.min(89.0f, pitch)); // Clamp pitch
        
        // Convert to quaternion
        rotation.identity()
            .rotateY((float) Math.toRadians(-yaw))
            .rotateX((float) Math.toRadians(-pitch));
        
        viewDirty = true;
    }
    
    /**
     * Sets the perspective projection parameters.
     * 
     * @param fov Field of view in degrees
     * @param aspectRatio Aspect ratio (width / height)
     * @param nearPlane Near clipping plane
     * @param farPlane Far clipping plane
     */
    public void setPerspective(float fov, float aspectRatio, float nearPlane, float farPlane) {
        this.fov = fov;
        this.aspectRatio = aspectRatio;
        this.nearPlane = nearPlane;
        this.farPlane = farPlane;
        projectionDirty = true;
    }
    
    /**
     * Enables or disables reverse-Z projection.
     * Reverse-Z provides better depth precision for large view distances.
     * Requires GL_ARB_clip_control support (OpenGL 4.5+).
     * 
     * @param useReverseZ Whether to use reverse-Z projection
     */
    public void setReverseZ(boolean useReverseZ) {
        if (this.useReverseZ != useReverseZ) {
            this.useReverseZ = useReverseZ;
            projectionDirty = true;
        }
    }
    
    /**
     * Sets up a look-at view matrix.
     * 
     * @param eye Camera position
     * @param target Target position to look at
     * @param up Up vector
     */
    public void lookAt(Vector3f eye, Vector3f target, Vector3f up) {
        this.position.set(eye);
        viewMatrix.setLookAt(eye, target, up);
        viewDirty = false;
    }
    
    /**
     * Moves the camera by a delta vector.
     * 
     * @param delta Movement delta
     */
    public void move(Vector3f delta) {
        position.add(delta);
        viewDirty = true;
    }
    
    /**
     * Rotates the camera by delta angles.
     * 
     * @param dyaw Yaw delta in degrees
     * @param dpitch Pitch delta in degrees
     */
    public void rotate(float dyaw, float dpitch) {
        setRotation(yaw + dyaw, pitch + dpitch);
    }
    
    /**
     * Updates the view matrix if dirty.
     */
    public void updateViewMatrix() {
        if (!viewDirty) return;
        
        // Create rotation matrix from quaternion
        Matrix4f rotationMatrix = new Matrix4f().set(rotation);
        
        // Create translation matrix
        Matrix4f translationMatrix = new Matrix4f().translation(-position.x, -position.y, -position.z);
        
        // Combine: view = rotation * translation
        viewMatrix.set(rotationMatrix).mul(translationMatrix);
        
        viewDirty = false;
    }
    
    /**
     * Updates the projection matrix if dirty.
     */
    public void updateProjectionMatrix() {
        if (!projectionDirty) return;
        
        float fovRadians = (float) Math.toRadians(fov);
        
        if (useReverseZ) {
            // Reverse-Z: swap near and far, use infinite far plane
            projectionMatrix.setPerspective(fovRadians, aspectRatio, farPlane, nearPlane);
        } else {
            // Standard projection
            projectionMatrix.setPerspective(fovRadians, aspectRatio, nearPlane, farPlane);
        }
        
        projectionDirty = false;
    }
    
    /**
     * Updates the combined view-projection matrix.
     */
    public void updateViewProjectionMatrix() {
        updateViewMatrix();
        updateProjectionMatrix();
        projectionMatrix.mul(viewMatrix, viewProjectionMatrix);
    }
    
    /**
     * Updates all matrices.
     */
    public void update() {
        updateViewProjectionMatrix();
    }
    
    /**
     * Gets the forward direction vector.
     * 
     * @return Forward vector
     */
    public Vector3f getForward() {
        return new Vector3f(0, 0, -1).rotate(rotation);
    }
    
    /**
     * Gets the right direction vector.
     * 
     * @return Right vector
     */
    public Vector3f getRight() {
        return new Vector3f(1, 0, 0).rotate(rotation);
    }
    
    /**
     * Gets the up direction vector.
     * 
     * @return Up vector
     */
    public Vector3f getUp() {
        return new Vector3f(0, 1, 0).rotate(rotation);
    }
    
    // Getters
    
    public Vector3f getPosition() {
        return new Vector3f(position);
    }
    
    public float getYaw() {
        return yaw;
    }
    
    public float getPitch() {
        return pitch;
    }
    
    public Matrix4f getViewMatrix() {
        updateViewMatrix();
        return viewMatrix;
    }
    
    public Matrix4f getProjectionMatrix() {
        updateProjectionMatrix();
        return projectionMatrix;
    }
    
    public Matrix4f getViewProjectionMatrix() {
        updateViewProjectionMatrix();
        return viewProjectionMatrix;
    }
    
    public float getFov() {
        return fov;
    }
    
    public float getAspectRatio() {
        return aspectRatio;
    }
    
    public float getNearPlane() {
        return nearPlane;
    }
    
    public float getFarPlane() {
        return farPlane;
    }
}
